[DEFEDIT-6399] Retain specific expensive outputs in editor cache

matgis opened the pull request:
Fixes #6399

### User-facing changes
* Reduced the time spent on subsequent Build and Save operations when editing a large project.
* Added a Shared Editor Settings editor to the Project menu where projects can adjust the cache size used by the editor for all users working in the project. If you still experience frequent delays in the editor after this change, try increasing the cache size.

### Technical changes
* The system cache now uses a custom LRU cache implementation.
* Certain expensive-to-compute outputs are protected from eviction from the cache as it fills. Effectively, these outputs do not count towards the cache limit, and will not be evicted to make room for new entries.
* To compensate, the default cache capacity to store the remaining outputs have been decreased from `40000` to `20000`.
* When creating a system cache, you can now provide a `:cache-retain?` predicate that takes an `Endpoint` and is expected to return `true` if the value should be retained in the cache even as it fills up.
* When loading a project, we will look for a file named `/project.shared_editor_settings`, and use its `performance.cache_capacity` setting for the size of the system cache, if specified.
* We now defer loading of namespaces related to the code view until after we've shown the Welcome screen.
* Added few helper functions to the `resource-node` namespace that lets us resolve the `Resource` from a resource node id without having to create an `evaluation-context`.
* Added `settings/SimpleSettingsResourceNode` to enable editing of simple ini-file format resources.
* `Endpoints` now implement the `Comparable` interface, enabling their use as keys in sorted maps.
* Evaluating a property on an `OverrideNode` no longer evaluates it on all originals recursively all the time.
* We now use transient updates when modifying the system cache, which is a bit more efficient for our use case.
* The JVM property `defold.tests` will now be set to `true` during tests.

matgis committed:
Projects can add a project.editor_tweaks file to tweak graph cache capacity

matgis committed:
Added failing editor test for save-data not being evicted when we exceed the cache limit

matgis committed:
Retain save-data in cache even after we exceed the cache limit

matgis committed:
Added editor test for build-targets not being evicted when we exceed the cache limit

matgis committed:
Minor optimizations

matgis committed:
Log cache metrics after save or build

matgis committed:
Reduce default editor cache limit for non-retained entries

matgis committed:
Added code for reconfiguring the system cache

matgis committed:
Renamed project.editor_tweaks file to project.shared_settings

matgis committed:
Add Shared Settings menu item to the Project menu

matgis committed:
Revert changes related to reconfiguring the editor cache at runtime, will be added back later

matgis committed:
Adjusted the wording on the Shared Settings `cache_capacity` tooltip

matgis commented on editor/src/clj/editor/defold_project.clj:
```diff
@@ -566,19 +569,18 @@
+          (g/transact transaction-metrics
+            (for [node-id (:mark-deleted plan)]
+              (let [flaw (resource-io/file-not-found-error node-id nil :fatal (resource-node/resource basis node-id))]
+                (g/mark-defective node-id flaw))))))
```
Mostly indentation due to obtaining the `basis` outside of the loop, but we're also using `resource-node/resource` here to avoid creating a bunch of `evaluation-contexts`.

matgis commented on editor/src/clj/editor/defold_project.clj:
```diff
@@ -839,25 +841,57 @@
+(defn- project-resource-node? [basis node-id]
+  (if-some [resource (resource-node/as-resource-original basis node-id)]
+    (some? (resource/proj-path resource))
+    false))
```
The questing we're asking is, "Does this node-id represent an actual resource in the project?". I.e. it's a file or a zip entry from a dependency. We only want to consider the original nodes here, since override nodes are created when instantiating components, etc. These override nodes present as the instantiated files in every other way, but they're not.

matgis committed:
Review fixes

matgis commented on editor/src/clj/editor/settings.clj:
```diff
@@ -13,18 +13,16 @@
-            [service.log :as log]))
+            [editor.resource-node :as resource-node]
+            [editor.settings-core :as settings-core]
+            [editor.workspace :as workspace]))
```
Mostly sorting. I guess I added `[editor.resource-node :as resource-node]` to the list.

matgis commented on editor/src/clj/internal/graph/types.clj:
```diff
@@ -28,6 +28,13 @@
+          node-id-comparison (Long/compare node-id (.-node-id that))]
+      (if (zero? node-id-comparison)
+        (compare label (.-label that))
+        node-id-comparison)))
```
I've added tests for this below.

matgis commented on editor/src/clj/internal/node.clj:
```diff
@@ -1544,7 +1544,9 @@
-    (get properties property (gt/get-property (gt/node-by-id-at basis original-id) basis property)))
+    (if-some [entry (find properties property)]
+      (val entry)
+      (gt/get-property (gt/node-by-id-at basis original-id) basis property)))
```
Evaluating properties on override nodes would always look up every node up the override chain and evaluate the property on each of them unnecessarily. Now we only recurse up the chain when our node doesn't provide an override value.

matgis commented on editor/test/integration/test_util.clj:
```diff
@@ -277,7 +277,8 @@
 (defn- load-system-and-project-raw [path]
-  (test-support/with-clean-system
+  (test-support/with-clean-system {:cache-size 1000
+                                   :cache-retain? project/cache-retain?}
```
Tests default to a cache size of `1000`, just like before. Increasing this doesn't appear to make things faster.

matgis commented on editor/test/internal/cache_test.clj:
```diff
@@ -13,18 +13,15 @@
-
-(defn- mock-system [cache-size ch]
-  (is/make-system {:cache-size cache-size}))
+            [support.test-support :refer :all]))
```
The `mock-system` function was unused. Every test was using `with-clean-system`.

matgis commented on editor/test/internal/cache_test.clj:
```diff
@@ -57,16 +54,171 @@
+            (is (= 0 (count cache)))
+            (is (unlimited-cache? cache))
+            (let [cache (populate-cache cache)]
+              (is (= 4 (count cache))))))))))
```
Note that `"retain rules remain after clear"` is tested in the `retain-test` below

matgis commented on editor/test/internal/node_test.clj:
```diff
@@ -609,3 +609,32 @@
+               (gt/endpoint array-input-producer-one :internal-output)
+               (gt/endpoint array-input-producer-two :evaluated-output)
+               (gt/endpoint array-input-producer-two :internal-output)}
+             (set (keys (g/cache))))))))
```
Added just to make sure caching works as expected for input dependencies, etc. There are probably lots of tests that would fail if these assumptions didn't hold, but I wanted something simple to run frequently while I was messing around with the caching.

matgis commented on editor/src/clj/editor/workspace.clj:
```diff
@@ -291,6 +290,7 @@ ordinary paths."
         (dialogs/make-info-dialog
           {:title "Unable to Load Plugin"
            :icon :icon/triangle-error
+           :always-on-top true
```
Snuck this fix in at the last minute. If there's an error in a plugin while loading the project, make sure the error dialog stays on top after the editor opens.

vlaaad commented:
This PR seems to be affecting our efforts on the #5447 front. I tried this code on `editor-dev` and this branch, using "lots_of_collections1.zip" repro in the linked issue:
```clj
(in-ns 'internal.graph)

(require 'criterium.core)

(let [c (dynamo.graph/cache)
      deps (basis-dependencies
             (dynamo.graph/now)
             #{#g/endpoint [(editor.defold-project/get-resource-node
                              (ffirst (dynamo.graph/targets-of 0 :resource-list))
                              "/main/main.script")
                            :modified-lines]})]
  (criterium.core/quick-bench
    (internal.cache/cache-invalidate c deps)))
```

Results on `editor-dev`:
```clj
;Evaluation count : 30 in 6 samples of 5 calls.
;             Execution time mean : 25,691161 ms
;    Execution time std-deviation : 4,889217 ms
;   Execution time lower quantile : 23,363918 ms ( 2,5%)
;   Execution time upper quantile : 34,175660 ms (97,5%)
;                   Overhead used : 7,127517 ns
;
;Found 1 outliers in 6 samples (16,6667 %)
;    low-severe     1 (16,6667 %)
; Variance from outliers : 48,2024 % Variance is moderately inflated by outliers
```
Results on this branch:
```clj
;Evaluation count : 18 in 6 samples of 3 calls.
;             Execution time mean : 42,450507 ms
;    Execution time std-deviation : 6,554726 ms
;   Execution time lower quantile : 36,502785 ms ( 2,5%)
;   Execution time upper quantile : 51,301903 ms (97,5%)
;                   Overhead used : 7,109459 ns
```
I'm worried this will hurt editor performance in a noticeable way...

vlaaad commented:
Maybe a different approach would be better, like adding a bespoke flag to `defnode` macro similar to `:cached` that will use a different type of cache altogether. I haven't thought this through, just spitballing here :)

matgis commented:
Ohh, nice find about the potential slowdown during cache invalidation! That's surprising. I'll look into it! We definitely do not make the cache-invalidation situation worse.

Regarding the approach taken, I considered the approach of tagging the affected outputs in the `defnode` declaration, but decided against it for multiple reasons. Let me give you some background.

The retain rules differ between output types, so it wouldn't be as simple as just tagging an output as `:cache-retained`. Save-related outputs are retained in the cache for nodes that represent actual files in the project, whereas build-related outputs are retained for both files and zip-entries. The node types would be the same for both files and zip-entries, so there would need to be something like `:cache :save` and `:cache :build` to categorize the outputs.

Unfortunately we cannot rely on the fact that only nodes representing actual project files (as opposed to zip-entries) are connected to the save system, because one of the outputs save relies on (and we want to retain), `:source-value`, is also evaluated in order to determine the load order of nodes through the `resource-node/resource-node-dependencies` function, which evaluates it on zip-entries as well as files, so we really do need the two different retain rules.

But even with that, we don't actually want too retain every output tagged `:cache :save` or `:cache :build`. A lot of the `:source-value` and `:build-targets` outputs are on override instances of resource nodes rather than the "real" resource nodes. Retaining everything would immensely grow the number of cached entries in a large project, so we only retain the value of these outputs for the "real" (i.e. original) resource nodes and take the hit of redoing the work in case a dependency changes. For non-originals, regular eviction rules apply.

One approach would be to refer to the `retain?` predicate in the output definition. Something like
```clj
(output save-data :cached :retain? retain-rules/save-related? ...)
(output source-value :cached :retain? retain-rules/save-related? ...)
(output build-targets :cached :retain? retain-rules/build-related? ...)
```
This looks like a promising approach, but it has some drawbacks. First, we'd have to modify every one of these outputs on every `defnode` that is also a `ResourceNode`, touching a lot of code and potentially missing a few outputs if we're not careful. Second, it introduces a dependency between every code file containing one of these node definitions to this new hypothetical `retain-rules` module and all its transitive dependencies, which might complicate things in the future.

I don't like involving philosophy when it comes to decisions about architecture because I think it tends to overshadow practical concerns, but you could make an argument for the decision around whether or not a value should be retained in the cache being a cache-concern, not a node-concern. Especially considering our system has multiple different cache implementations, and both the null cache and unlimited cache implementations aren't concerned about retaining entires at all.

I hope this explains my reasoning for the approach I've taken, but I love to discuss it further if we can do better!

matgis committed:
Use transient collections in system cache implementations

matgis committed:
Stop using original LRUCache implementation

matgis committed:
Reduce code duplication between many and single cache miss

matgis commented:
I was not able to reproduce as drastic of a slowdown as you were seeing. Perhaps there was still some background processing going on when you ran the test on my branch? I was also surprised to see my times being so much lower than yours. However, I did see a slight but consistent degradation of performance on my branch compared to `editor-dev`. On my branch, evicting these elements from the cache had around 113% of the runtime compared to `editor-dev`.

Attempting to isolate the cause of the slowdown, I was getting inconsistent run times. Even when I used the same cache implementation as on `editor-dev`, it seems the mere presence of my additional code around it was causing it to run slower. The only way I could get it back to running at the original speed was to remove my extension protocol from the cache implementations. I tried moving the functionality to multimethods instead, or even to standalone functions switching on class internally. Nothing worked. It was as if the mere presence of my code was causing calls to `core.cache/evict` to become slower.

Eventually I gave up and concluded my change was causing the generated bytecode to exceed some kind of internal JIT-limit, changing how the code was being optimized internally. I decided then to try to minimize the number of calls to the `core.cache/evict` function. The pattern used in the editor is to evict many cache entries at once, so I tried to use transient updates of the cache internals instead. Unfortunately, the `data.priority-map` used for the LRU does not support transient updates, so only the cache map itself benefits from this. After this change, the cache evictions surpassed their original speed on `editor-dev`. On my machine I was seeing around 70% of the runtime compared to `editor-dev`.

### My machine running editor-dev (baseline)
```
Number of dependencies: 40535
WARNING: Final GC required 30.7771517470939 % of runtime
Evaluation count : 192 in 6 samples of 32 calls.
             Execution time mean : 3,214822 ms
    Execution time std-deviation : 48,177338 µs
   Execution time lower quantile : 3,179466 ms ( 2,5%)
   Execution time upper quantile : 3,291455 ms (97,5%)
                   Overhead used : 6,830577 ns
```

### My branch before addressing slowdown
```
Number of dependencies: 40535
WARNING: Final GC required 30.34009822798207 % of runtime
Evaluation count : 168 in 6 samples of 28 calls.
             Execution time mean : 3,623864 ms
    Execution time std-deviation : 42,324257 µs
   Execution time lower quantile : 3,594355 ms ( 2,5%)
   Execution time upper quantile : 3,685611 ms (97,5%)
                   Overhead used : 6,830619 ns
```

### My branch after using transient cache updates
```
Number of dependencies: 40535
WARNING: Final GC required 30.1312520559907 % of runtime
Evaluation count : 270 in 6 samples of 45 calls.
             Execution time mean : 2,247145 ms
    Execution time std-deviation : 15,017086 µs
   Execution time lower quantile : 2,231313 ms ( 2,5%)
   Execution time upper quantile : 2,268298 ms (97,5%)
                   Overhead used : 6,832376 ns
```

matgis commented:
Here's the code I used to perform the measurements in isolation without loading up the editor:
```clj
(ns scratch
  (:require [criterium.core :as criterium]
            [dynamo.graph :as g]
            [editor.defold-project :as project]
            [editor.editor-extensions :as extensions]
            [editor.progress :as progress]
            [editor.workspace :as workspace]
            [integration.test-util :as tu]
            [internal.cache :as c]
            [internal.graph.types :as gt]
            [support.test-support :as ts]))

(let [project-dir-path "/Users/mats/Defold/lot_of_collections1"]
  (System/gc)
  (ts/with-clean-system
    {:cache-size 20000
     :cache-retain? project/cache-retain?}
    (let [workspace (tu/setup-workspace! world project-dir-path)
          game-project-resource (workspace/find-resource workspace "/game.project")
          main-script-resource (workspace/find-resource workspace "/main/main.script")
          project-graph (g/make-graph! :history true :volatility 1)
          extensions (extensions/make project-graph)
          project (project/open-project! project-graph extensions workspace game-project-resource progress/null-render-progress!)]
      (let [main-script (project/get-resource-node project main-script-resource)
            basis (g/now)
            cache (g/cache)
            dependencies (gt/dependencies basis [(gt/endpoint main-script :modified-lines)])]
        (println "Number of dependencies:" (count dependencies))
        (criterium/quick-bench
          (c/cache-invalidate cache dependencies))))))
```

matgis committed:
Renamed `.shared_settings` extension to `.shared_editor_settings`

vlaaad commented:
I have a different number of dependencies (163454), I changed the project at some point to have even more references so the issue is more noticeable.

vlaaad commented on editor/test/integration/app_view_test.clj:
```diff
@@ -166,3 +169,55 @@
+        cache-size 50
+        retained-labels #{:build-targets}]
+    (letfn [(cached-endpoints []
+              (into (sorted-set)
```
I'm not sure we need to make endpoints comparable and use a sorted set here. From what I see, it's used for `subset?` and `contains?` check that don't need set ordering. Or are there other uses that I missed?

vlaaad commented on editor/src/clj/editor/shared_editor_settings.clj:
```diff
@@ -0,0 +1,109 @@
+                system-config-or-exception))))
+        default-system-config)))
+
+;; Called through reflection.
```
I guess you can call ns/var resolution a reflection :D

vlaaad reviewed (approved):
LGTM! I added some questions and comments. Not sure we need the Endpoint class to be comparable. Very good that you added many cache tests 👍 

I don't think a place to specify caching rules is a philosophical architecture issue, I think caching is pretty hard and sensitive and requires fine-tuning for best results, and our approach of simply marking outputs as `:cached` is oversimplifying things too much. It would be good to have more control in specifying caching behavior/scopes. I find the solution in this PR better than what was before, so I think we can merge it. 

I would still prefer us to spend some time afterwards thinking/designing caching behavior. I like the idea of specifying retain predicate on the defnode definition, even if it will require us modifying all affected outputs — I think this explicitness is good. Maybe make an issue so we don't forget about it?

matgis committed:
Review fixes

Don't call g/now for every retained key when encaching items

Expose endpoint-accessor macros to dynamo.graph namespace

Review fixes

vlaaad reviewed (approved):
LGTM! I agree that Endpoint implementing Comparable doesn't hurt :)

matgis committed:
Merge branch 'editor-dev' into DEFEDIT-6399-slow-save

matgis committed:
Don't try to load missing spine documentation at editor startup
